tenant/src/main/scala/com/improving/app/tenant/domain/Tenant.scala

```diff
@@ -0,0 +1,391 @@
+        state match {
+          case x: DraftTenant => x.copy(info = x.info.copy(address = value.newAddress), metaInfo = value.metaInfo.get)
+          case x: ActiveTenant => x.copy(info = x.info.copy(address = value.newAddress), metaInfo = value.metaInfo.get)
+          case x: SuspendedTenant => x.copy(info = x.info.copy(address = value.newAddress), metaInfo = value.metaInfo.get)
+        }
+      case TenantEventMessage.SealedValue.TenantActivatedValue(value) =>
+        state match {
+          case x: DraftTenant => ActiveTenant(x.info, value.metaInfo.get)
+          case _: ActiveTenant => state // tenant cannot have TenantActivated in Active state
+          case x: SuspendedTenant => ActiveTenant(x.info, value.metaInfo.get)
+        }
+      case TenantEventMessage.SealedValue.TenantSuspendedValue(value) =>
+        state match {
+          case x: DraftTenant => SuspendedTenant(x.info, value.metaInfo.get, value.suspensionReason)
+          case x: ActiveTenant => SuspendedTenant(x.info, value.metaInfo.get, value.suspensionReason)
+          case x: SuspendedTenant => SuspendedTenant(x.info, value.metaInfo.get, value.suspensionReason)
+        }
+    }
+  }
+
+  /**
+   * Updates the meta info which includes updating the lastUpdated timestamp
+   * @param metaInfo
+   * @param lastUpdatedByOpt
+   * @return
+   */
+  private def updateMetaInfo(metaInfo: MetaInfo, lastUpdatedByOpt: Option[MemberId]): MetaInfo = {
+    val currentTime = Timestamp(Instant.now())
+    metaInfo.copy(lastUpdatedBy = lastUpdatedByOpt, lastUpdated = Some(currentTime))
+  }
+
+  /**
+   * Validation and business logic for updating the tenant name
+   * @param info
+   * @param metaInfo
+   * @param updateTenantNameCommand
+   * @param replyTo
+   * @return
+   */
+  private def updateTenantNameLogic(
+    info: Info, metaInfo: MetaInfo,
+    updateTenantNameCommand: UpdateTenantName,
+    replyTo: ActorRef[StatusReply[TenantEvent]]
+  ): ReplyEffect[TenantEvent, TenantState] = {
+    if (info.name.equals(updateTenantNameCommand.newName)) {
+      Effect.reply(replyTo)(
+        StatusReply.Error("Tenant name is already in use")
+      )
+    } else {
+      val newMetaInfo = updateMetaInfo(metaInfo = metaInfo, lastUpdatedByOpt = updateTenantNameCommand.updatingUser)
+      val event = TenantNameUpdated(
+        tenantId = updateTenantNameCommand.tenantId,
+        oldName = info.name,
+        newName = updateTenantNameCommand.newName,
+        metaInfo = Some(newMetaInfo)
+      )
+
+      Effect.persist(event).thenReply(replyTo) { _ => StatusReply.Success(event) }
+    }
+  }
+
+  /**
+   * State business logic for updating the tenant name
+   * @param state
+   * @param updateTenantNameCommand
+   * @param replyTo
+   * @return
+   */
+  private def updateTenantName(
+    state: TenantState,
+    updateTenantNameCommand: UpdateTenantName,
+    replyTo: ActorRef[StatusReply[TenantEvent]]
+  ): ReplyEffect[TenantEvent, TenantState] = {
+    state match {
```

Comment:
This is a command handler. UpdateTenantName could come from internal or external sources. You need to add require statements or some other kind of input validation here so we don't do silly things like update the tenant name to an empty string, for example.  Same is true of all other command processors. These are the same idea as preconditions found in the Eiffel programming language to support "Program By Contract" .. by Bertrand Meyer.

Reply:
Went ahead and started using a more "contract programming" approach to the tenant actor.